File: vocabtivator-sentence/src/main/java/org/bubbasmith/vocabtivator/integration/external/ContextReversoSentencesSearch.java
package org.bubbasmith.vocabtivator.integration.external;

import org.bubbasmith.vocabtivator.integration.dao.SentenceRepository;
import org.bubbasmith.vocabtivator.model.Sentence;
import org.bubbasmith.vocabtivator.integration.parser.ContextReversoHtmlParser;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;
import org.springframework.web.reactive.function.client.WebClient;
import reactor.core.publisher.Mono;

import java.net.URI;
import java.util.List;

@Service
public class ContextReversoSentencesSearch implements ExternalSentencesSearch {

    private final static URI SEARCH_URL = URI.create("https://context.reverso.net/translation/english-russian/");

    private final ContextReversoHtmlParser contextReversoHtmlParser;
    private final WebClient webClient;

    public ContextReversoSentencesSearch(ContextReversoHtmlParser contextReversoHtmlParser,
                                         @Qualifier("contextReversoWebClient") WebClient webClient) {
        this.contextReversoHtmlParser = contextReversoHtmlParser;
        this.webClient = webClient;
    }

    public Mono<List<Sentence>> findSentencesForWord(String word) {
        return webClient.get().uri(SEARCH_URL + word.replaceAll(" ", "+"))
                .header(HttpHeaders.CONTENT_TYPE, MediaType.TEXT_HTML_VALUE)
                .exchange()
                .flatMap(it -> it.bodyToMono(String.class))
                .map(contextReversoHtmlParser::parse);
    }
}

File: vocabtivator-sentence/src/main/java/org/bubbasmith/vocabtivator/integration/external/ContextReversoSentencesSearch.kt
package org.bubbasmith.vocabtivator.integration.external

import org.bubbasmith.vocabtivator.integration.parser.ContextReversoHtmlParser
import org.bubbasmith.vocabtivator.model.Sentence
import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.http.HttpHeaders
import org.springframework.http.MediaType
import org.springframework.stereotype.Service
import org.springframework.web.reactive.function.client.WebClient
import org.springframework.web.reactive.function.client.awaitBody
import org.springframework.web.reactive.function.client.awaitExchange
import java.net.URI

@Service
class ContextReversoSentencesSearch(
        private val contextReversoHtmlParser: ContextReversoHtmlParser,
        @Qualifier("contextReversoWebClient")  private val webClient: WebClient): ExternalSentencesSearch {

    private val searchUrl: URI = URI.create("https://context.reverso.net/translation/english-russian/")
    
    override suspend fun findSentencesForWord(word: String): List<Sentence> {
        val body = webClient.get().uri(searchUrl.toString() + word.replace(" ".toRegex(), "+"))
                .header(HttpHeaders.CONTENT_TYPE, MediaType.TEXT_HTML_VALUE)
                .awaitExchange()
                .awaitBody<String>()
        return contextReversoHtmlParser.parse(body)

    }
}